hera_mc/geo_handling.py

```diff
@@ -313,33 +322,292 @@ def get_location(self, to_find, query_date, show_location=False, verbosity='m'):
+            elif at_date > now:
+                loop = False
+        return fc, at_date
+
+    def get_all_fully_connected_ever(self, earliest_date=Time('2016-09-01'),
+                                     full_req=part_connect.full_connection_parts_paper,
+                                     station_types_to_check='all'):
         """
+        Returns a list of all of the locations fully connected on at_date that
+        have station_types in station_types_to_check.
+
+        Each location is returned as one dict in the list. Dict keys are:
+            'station_name': name of station (string, e.g. 'HH27')
+            'station_type': type of station (string, e.g. 'HH', 'PI', etc)
+            'tile': UTM tile name (string, e.g. '34J'
+            'datum': UTM datum (string, e.g. 'WGS84')
+            'easting': station UTM easting (float)
+            'northing': station UTM northing (float)
+            'longitude': station longitude (float)
+            'latitude': station latitude (float)
+            'elevation': station elevation (float)
+            'antenna_number': antenna number (integer)
+            'correlator_input_x': correlator input for x (East) pol (string e.g. 'DF7B4')
+            'correlator_input_y': correlator input for y (North) pol (string e.g. 'DF7B4')
+            'start_date': start of connection in gps seconds (long)
+            'stop_date': end of connection in gps seconds (long or None no end time)
 
-        stations = self.session.query(geo_location.GeoLocation).all()
-        connections = self.session.query(part_connect.Connections).all()
-        stations_conn = []
-        for stn in stations:
-            hera_proj = Proj(proj='utm', zone=stn.tile, ellps=stn.datum, south=True)
-            stn.lon, stn.lat = hera_proj(stn.easting, stn.northing, inverse=True)
-            ever_connected = self.is_in_connections(stn.station_name)
-            if ever_connected:
-                connections = self.session.query(part_connect.Connections).filter(
-                    part_connect.Connections.upstream_part == stn.station_name)
-                for conn in connections:
-                    ant_num = int(conn.downstream_part[1:])
+        Parameters
+        -----------
+        earliest_date:  earliest_date to check for connections.
+        full_req: list of parts required for a connection to be a 'full' connection
+            (default: part_connect.full_connection_parts_paper)
+        station_types_to_check:  list of station types to limit check, or all (default: all)
+        """
+        from hera_mc import cm_hookup
+        from sqlalchemy import asc
+        hookup = cm_hookup.Hookup(self.session)
+
+        base_tdelta = 86400.0
+        m = 10.0
+        self.get_station_types()
+        station_conn = []
+        now = cm_utils._get_astropytime('now')
+        for k, stn_type in self.station_types.iteritems():
+            if station_types_to_check == 'all' or k in station_types_to_check:
```

Comment:
I think the logic here should be inverted to reduce the amount of rightward drift in the code:

```python
if station_types_to_check != 'all' and k not in station_types_to_check:
    continue

for stn in ...
```